models/intermediate/int_transform_contract_code.sql

```diff
@@ -0,0 +1,175 @@
+{{
+    config(
+        materialized = 'table',
+        unique_key = ['ledger_key_hash', 'closed_at'],
+        cluster_by = ["ledger_key_hash", "closed_at", "row_hash"],
+        tags = ["soroban_analytics", "intermediate", "daily"]
+    )
+}}
+
+/*
+Model: int_transform_contract_code
+
+Description:
+------------
+This intermediate model combines incremental loading and transformations for contract code data.
+It processes new or updated data, calculates new fields, and performs deduplication.
+
+Load Type: Truncate and Reload
+
+Features:
+---------
+1. Uses execution_date from Airflow to determine the processing window.
+2. Fetches data till the previous day from the Staging layer, one key hash per day (latest).
+3. Calculates `contract_create_ts` and `contract_delete_ts` from staging data.
+4. Joins the calculated timestamps with the incremental data and applies transformations.
+5. Calculates a row hash using SHA256 on all relevant fields.
+6. Performs deduplication based on the row hash of the fully transformed data. This step primarily detects duplicates in consecutive rows.
+
+Usage:
+------
+Compile:
+    dbt compile --models int_transform_contract_code
+
+Run:
+    dbt run --models int_transform_contract_code --full-refresh
+    dbt run --models int_transform_contract_code
+    dbt run --models int_transform_contract_code --vars '{"execution_date": "2024-10-01T00:00:00+00:00"}'
+*/
+
+-- Set the execution date (use Airflow value, fallback to dbt default if absent)
+{% set execution_date = var('execution_date', dbt_airflow_macros.ts()) %}
+
+with
+    source_data as (
+        select
+            contract_code_hash
+            , ledger_key_hash
+            , last_modified_ledger
+            , ledger_entry_change
+            , ledger_sequence
+            , deleted
+            , closed_at
+            , n_instructions
+            , n_functions
+            , n_globals
+            , n_table_entries
+            , n_types
+            , n_data_segments
+            , n_elem_segments
+            , n_imports
+            , n_exports
+            , n_data_segment_bytes
+            , batch_id
+            , batch_run_date
+            , cast('{{ execution_date }}' as timestamp) as airflow_start_ts
+            -- Rank records to identify the latest entry
+            , row_number() over (
+                partition by ledger_key_hash, cast(closed_at as date)
+                order by closed_at desc, ledger_sequence desc
+            ) as row_num
+        from {{ ref('stg_contract_code') }}
+        where
+            -- Process only completed days up to execution_date
+            date(closed_at) < date('{{ execution_date }}')
```

Comment:
What's the reason to query beginning of time to execution date?

Reply:
- Originally had an Incremental Layer with Incremental Materialization and then a Truncate & Reload Type Transformation Layer. But had to combine them into one Model to prevent "[Rejoining of Upstream Concepts](https://dbt-labs.github.io/dbt-project-evaluator/latest/rules/modeling/#rejoining-of-upstream-concepts)" issue

- After combining them, if I make this Model incremental, Intra Dupes can come in as Incremental Model will only look at newer versions of data based of closed_at.
Ex: If i have a ledger: 123 and $100 with closed_date : 10/28 loaded today and another record comes through with the same data on 10/29 , Incremental Model allows this Dupe to get in and the De-Duplication logic i have in here to prevent consecutive row dupes will not detect that.

- Considering all this and the volumes, when I merged the two Models, made this model and full truncate and reload every day.

Reply:
I think my final comment applies to this question

> I think the main issue is that the current models do full table scans on every run int_* is essentially a full_refresh on every run --> dim_* are incremental. I think the int_* logic should also be incremental

There should be a way to process scd2 tables without doing a full table scan on stg_* data

Reply:
when you say process, does it mean just reprocessing existing data in the table cause some bad date came in and disrupted the chaining? If yes, I am working on a stored proc for now to fix data within the table. I didn't include such things for MVP.

But the long term plan though is to have a framework covering all these scenarios -

-         Initial Load
-         Daily Incremental Run - Insert (Future Dated Transaction)
-         Daily Incremental Run - No Change (Future Dated Transaction)
-         Daily Incremental Run - Update/NoChange (Future Dated Transaction as well as past dated transaction)
-         Multi-Day Catchup (All Future Dated Transactions with or without change in data)
-         Single-Day Recovery (past dated transaction with or without data change)
-         Multi-Day Recovery (past dated transactions with or without data change)
-         Full Recovery (Full Wipe)
-         Late Arriving records